src/lib/transition_frontier/transition_frontier0.ml

```diff
@@ -0,0 +1,645 @@
+   *      best tip node
+   *   3) set the new node as the best tip if the new node has a greater length than
+   *      the current best tip
+   *   4) move the root if the path to the new node is longer than the max length
+   *       I   ) find the immediate successor of the old root in the path to the
+   *             longest node (the heir)
+   *       II  ) find all successors of the other immediate successors of the
+   *             old root (bads)
+   *       III ) cleanup bad node masks, but don't garbage collect yet
+   *       IV  ) move_root the breadcrumbs (rewires staged ledgers, cleans up heir)
+   *       V   ) garbage collect the bads
+   *       VI  ) grab the new root staged ledger
+   *       VII ) notify the consensus mechanism of the new root
+   *       VIII) if commit on an heir node that just emitted proof txns then
+   *             write them to snarked ledger
+  *)
+  let add_breadcrumb_exn t breadcrumb =
+    O1trace.measure "add_breadcrumb" (fun () ->
+        let hash =
+          With_hash.hash (Breadcrumb.transition_with_hash breadcrumb)
+        in
+        let root_node = Hashtbl.find_exn t.table t.root in
+        (* 1 *)
+        attach_breadcrumb_exn t breadcrumb ;
+        let node = Hashtbl.find_exn t.table hash in
+        (* 2 *)
+        let distance_to_parent = node.length - root_node.length in
+        let best_tip_node = Hashtbl.find_exn t.table t.best_tip in
+        (* 3 *)
+        if node.length > best_tip_node.length then t.best_tip <- hash ;
+        (* 4 *)
+        let maybe_new_root =
+          if distance_to_parent > max_length then (
+            Logger.info t.logger
+              !"Distance to parent: %d exceeded max_lenth %d"
+              distance_to_parent max_length ;
+            (* 4.I *)
+            let heir_hash = List.hd_exn (hash_path t node.breadcrumb) in
+            let heir_node = Hashtbl.find_exn t.table heir_hash in
+            (* 4.II *)
+            let bad_hashes =
+              List.filter root_node.successor_hashes
+                ~f:(Fn.compose not (State_hash.equal heir_hash))
+            in
+            let bad_nodes =
+              List.map bad_hashes ~f:(Hashtbl.find_exn t.table)
+            in
+            (* 4.III *)
+            let root_staged_ledger =
+              Breadcrumb.staged_ledger root_node.breadcrumb
+            in
+            let root_ledger = Inputs.Staged_ledger.ledger root_staged_ledger in
+            List.map bad_nodes ~f:breadcrumb_of_node
+            |> List.iter ~f:(fun bad ->
+                   ignore
+                     (Ledger.unregister_mask_exn root_ledger
+                        ( Breadcrumb.staged_ledger bad
+                        |> Inputs.Staged_ledger.ledger )) ) ;
+            (* 4.IV *)
+            let new_root_node = move_root t heir_node in
+            (* 4.V *)
+            let garbage = List.bind bad_hashes ~f:(successor_hashes_rec t) in
+            List.iter garbage ~f:(Hashtbl.remove t.table) ;
```

Comment:
we "garbage collect" the garbage here, so you won't be able to grab them from the hashtable as you did below -- instead, maybe you should grab the garbage breadcrumbs here